text/0028-ed25519-validation/0028-ed25519-validation.md

```diff
@@ -0,0 +1,79 @@
+- decoding of elliptic curve points as described in [Section 5.1.3](https://tools.ietf.org/html/rfc8032#section-5.1.3)
+- validation itself as described in [Section 5.1.7](https://tools.ietf.org/html/rfc8032#section-5.1.7)
+
+# Motivation
+
+Based on [research done by Henry de Valence](https://hdevalence.ca/blog/2020-10-04-its-25519am) we know that: 
+1. Not all implementations follow the decoding rules defined in RFC 8032, but instead accept non-canonically encoded inputs.
+2. The Ed25519 RFC provides two alternative verification equations, whereas one is stronger than the other. Different implementations use different equations and therefore validation results vary even across implementations that follow the RFC 8032. 
+
+This lack of consistent validation behavior is especially critical for IOTA as they can cause a breach of consensus across node implementations! For example, one node implementation may consider a particular transaction valid and mutate the ledger state accordingly, while a different implementation may discard the same transaction due to invalidity. This would result in a network fork and could only be resolved outside of the protocol. Therefore, an explicit and unambiguous definition of validation criteria, such as ZIP-215, is necessary.
+
+Furthermore, it is important to note that the holder of the secret key can produce more than one valid distinct signature. Such transactions with the same essence but different signatures are considered as double spends by the consensus protocol and handled accordingly. While this does not pose a problem for the core protocol, it may be a problem for 2nd layer solutions, similar to how [transaction malleability in bitcoin presented an issue for the lightning network](https://en.bitcoinwiki.org/wiki/Transaction_Malleability#How_Does_Transaction_Malleability_Affect_The_Lightning_Network.3F).
+
+# Detailed Design
+
+In order to have consistent validation of Ed25519 signatures for all edge cases and throughout different implementations, this RFC proposes explicit validation criteria. These three criteria **must** be checked to evaluate whether a signature is valid.
+
+Using the notation and Ed25519 parameters as described in the RFC 8032, the criteria are defined as follows:
+
+1. Accept non-canonical encodings of A and R.
+2. Reject values for S that are greater or equal than L.
+3. Use the equation [8][S]B = [8]R + [8][k]A' for validation.
+
+In the following, we will explain each of these in more detail.
+
+## Decoding
+
+The Curve25519 is defined over the finite field of order p=2<sup>255</sup>−19. A curve point (x,y) is encoded into its compressed 32-byte representation, namely by the 255-bit encoding of the field element y followed by a single sign bit that is 1 for negative x and 0 otherwise. This approach provides a unique encoding for each valid point. However, there are two classes of edge cases representing non-canonical encodings of valid points:
```

Comment:
You are obviously correct that field elements do not have a sing. However, my reasoning was the following:
- The Ed25519 paper contains the following definition: _x is negative if the (...) encoding of x is lexicographically larger than the (...) encoding of −x (...) the negative elements of F_q are {1,3,5,...,q−2}._ This is obviously well-defined in the sense that if x is not negative then -x mod q is negative.
- The RFC8032 itself says: _An element (x,y) of E is encoded as a b-bit string called ENC(x,y), which is the (b-1)-bit encoding of y concatenated with one bit that is 1 if x is negative and 0 if x is not negative._

That is why I thought talking about "negative" is more intuitive while still being formally correct. What do you suggest to improve this?

Reply:
The definition of a "negative" field element is not intuitive, especially when it really means `x (mod p) (mod 2)`. Let's maybe just add a reference to RFC8032:

```
sign bit that is 1 for negative x (as per RFC 8032, section 3.2) and 0 otherwise
```

Reply:
I've added this remark to the RFC.